designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/ActionService.scala

```diff
@@ -181,9 +190,62 @@ class ActionService(
       dbioRunner.runInTransaction(actionRepository.withLockedTable(dbioAction))
     }
 
+    private def fetchAndUpdateScenarioIfNeeded(processId: ProcessIdWithName)(implicit user: LoggedUser) = {
+      scenarioGraphSource match {
+        case LatestVersion =>
+          processRepository.fetchLatestProcessDetailsForProcessId[ScenarioDetailsShape](processId.id)
+        case FromGraph(scenarioGraph, scenarioLabels, baseScenarioVersionId) =>
+          processRepository
+            .fetchProcessDetailsForId[CanonicalProcess](processId.id, baseScenarioVersionId)
+            .flatMap {
+              case Some(scenario) if CanonicalProcessConverter.toScenarioGraph(scenario.json) == scenarioGraph =>
+                val scenarioWithTargetShape =
+                  scenario.mapScenario(ScenarioShapeFetchStrategy.convertToTargetShape[ScenarioDetailsShape])
+                DBIOAction.successful(Some(scenarioWithTargetShape))
+              case Some(scenario) =>
+                updateScenario(processId, scenarioGraph, scenarioLabels, scenario)
+              case None =>
+                DBIOAction.successful[Option[ScenarioWithDetailsEntity[ScenarioDetailsShape]]](None)
+            }
+      }
+    }
+
+    private def updateScenario(
+        processId: ProcessIdWithName,
+        scenarioGraph: ScenarioGraph,
+        scenarioLabels: Option[List[ProcessingType]],
+        scenario: ScenarioWithDetailsEntity[CanonicalProcess]
+    )(implicit user: LoggedUser) = {
+      val parameters = scenarioParametersServiceProvider.combined
+        .getParametersWithReadPermissionUnsafe(scenario.processingType)
+      val scenarioDetails = ScenarioWithDetailsConversions
+        .fromEntityIgnoringGraphAndValidationResult(scenario, parameters)
+      val updateCommand = UpdateScenarioCommand(scenarioGraph, None, scenarioLabels)
+      processService
+        .updateProcessDBIO(processId, scenarioDetails, updateCommand)
+        .flatMap(response =>
+          response.processResponse match {
```

Comment:
1. It looks strange that in our domain service (`ActionService`) we process some DTO. I see that it is a deeper problem, because `ProcessService` already is not a real domain service and it returns DTOs in other methods. Do you see some easy way to fix it, or is it too much work to tidy it up?
2. Also, this is very tricky that some `processResponse: None` means that nothing changed in the version. In scala, I would expect:
- ADT: `NoVersionChanges` | `NewVerssionWasAdded(versionId)`
- Nested class with a field called `newVersionInformation: Option[VersionId]`
-  Single `Option[VersionId]` as a return type with some comment

Why this method returns scenario id and scenario name that were passed?